packages/remix-react/single-fetch.ts

```diff
@@ -0,0 +1,203 @@
+      // 1. we only call the server if at least one of them calls `serverLoader`
+      // 2. if multiple call `serverLoader` only one fetch call is made
+      let singleFetchPromise: Promise<SingleFetchResults>;
+
+      let makeSingleFetchCall = async () => {
+        let url = singleFetchUrl(
+          addRevalidationParam(
+            manifest,
+            routeModules,
+            matches,
+            // Single fetch doesn't need/want naked index queries on action
+            // revalidation requests
+            stripIndexParam(request.url)
+          )
+        );
+
+        let result = await fetchAndDecode(url);
+        return result as SingleFetchResults;
+      };
+
+      singleFetch = async (routeId) => {
+        if (!singleFetchPromise) {
+          singleFetchPromise = makeSingleFetchCall();
+        }
+        let results = await singleFetchPromise;
+        if (results[routeId] !== undefined) {
+          return unwrapSingleFetchResult(results[routeId], routeId);
+        }
+        return null;
+      };
+    }
+
+    // Call the route handlers passing through the `singleFetch` function that will
+    // be called instead of making a server call
+    return Promise.all(
+      matches.map(async (m) => {
+        return m.resolve((handler) => handler(() => singleFetch(m.route.id)));
+      })
+    );
+  };
+}
+
+function stripIndexParam(_url: string) {
+  let url = new URL(_url);
+  let indexValues = url.searchParams.getAll("index");
+  url.searchParams.delete("index");
+  let indexValuesToKeep = [];
+  for (let indexValue of indexValues) {
+    if (indexValue) {
+      indexValuesToKeep.push(indexValue);
+    }
+  }
+  for (let toKeep of indexValuesToKeep) {
+    url.searchParams.append("index", toKeep);
+  }
+
+  return url.href;
+}
+
+// Determine which routes we want to load so we can add a `?_routes` search param
+// for fine-grained revalidation if necessary.  If a route has not yet been loaded
+// via `route.lazy` then we know we want to load it because it's by definition a
+// net-new route.  If it has been loaded then `shouldLoad` will have taken
+// `shouldRevalidate` into consideration.
+//
+// There is a small edge case that _may_ result in a server loader running
+// _somewhat_ unintended, but it's unavoidable:
+// - Assume we have 2 routes, parent and child
+// - Both have `clientLoader`'s and both need to be revalidated
+// - If neither calls `serverLoader`, we won't make the single fetch call
+// - We delay the single fetch call until the **first** one calls `serverLoader`
+// - However, we cannot wait around to know if the other one calls
+//   `serverLoader`, so we include both of them in the `X-Remix-Routes`
+//   header
+// - This means it's technically possible that the second route never calls
+//   `serverLoader` and we never read the response of that route from the
+//   single fetch call, and thus executing that `loader` on the server was
+//   unnecessary.
+function addRevalidationParam(
```

Comment:
This method handles our fine-grained revalidation query param.  This should only show up on `.data` requests if the app is has implemented shouldRevalidate on one or more of the `matches`.  Otherwise, we don't include a query param and while we may run a few extra loaders, the cacheability of the request should improve greatly resulting in a net-positive impact on server load.